### ISSUE TITLE
Dolibarr is using 2005 PHP code, in 2018

### ISSUE THREAD

**Issue by ChangePlaces** [Dolibarr/dolibarr#10054]
I'm completely new to Dolibarr, I've just spent some time looking at the core code of 8. There are a number of _major_ issues which render Dolibarr unusable and unmaintable in 2018. This is not a "modern" ERP / CRM.

- SQL code and template code and logic are thrown together. This is a fundametal no-no since... about 2005
- Use a template system, They're a big thing for a reason. They need to be used. This will allow people to develop skins 100x more easily, which will put more themes in the marketplace etc
- As horrid as it may seem to French people, English is the global language. Directories, comments and variable names need to use it as you're not coding for yourself. Even dolistore should have English as the first language.
- There are many problems that have already been solved by external libraries, this is what composer is for
- Not having any symfony components is a bit of an issue - all worthy PHP projects are using Symfony components because they work. Not to be cool. They have a large community base actively working on solving specific problems and trying to solve them yourself just isn't cool anymore.
- Modules need a separate directory to be put in. I don't think anyone wants to copy third-party code into the project root due to malicious code overwriting core system files. 
- A router is needed so all code is handled from one file, to stop code duplication and to enable controllers to handle grouped functionality in one place (making the codebase instantly understandable)

Yes, this will need major refactoring, I'm sure it'll also mean people will have to learn new skills, but to be honest, Dolibarr is not a modern codebase.  I'd love to help contribute to Dolibarr, but I simply can't advocate putting code in the htdocs, something I haven't done since about 2005

I understand I'm going to be shot down for this, probably even have this closed / deleted but I have written this and have taken time to contribute to the Dolibarr project. I would like to contribute though, just not to the current spaghetti code

**Comment by torvista** [comment]
No doubt the developers are fully aware of modern methods and don't need someone to explain what is needed.
Dolibarr is what it is. A legacy project with a long history and the inevitable historical baggage that comes with it's roots as a french project. Major architecture changes mean major disruption for users on upgrades and so must be managed very slowly.
As far as I have seen, most users are really just users and do not modify much, judging by the few contributors here. So, architecture changes that don't improve functionality would seem pointlessly disruptive.
Dolibarr is not perfect by a very long way, but as far as I have found, it has greater functionality than similar projects, and if it doesn't, I can modify it relatively easily with my limited php knowledge.
In the end, the functionality is what matters to 100% of the people. If it works but is messy underneath the hood, that matters only to the very small percentage who need to go there, me included.

The facts are that in 2018 Dolibarr **is** being used and **is** being maintained, so that detracts from your otherwise valid comment.

I would have preferred to invest my time and contributions in something more modern too, but my skills are limited and I have found nothing better that worked out-of-the-box.

**Comment by mwelters** [comment]
@torvista Totally agree ! 
I also searched other projects for usability and decided for dolibarr BECAUSE it is functional and the code is easy to understand.

@ChangePlaces This is Opensource ! 
Fork it and show us a way to the enlightment of object-oriented modern component based Programming  !

**Comment by dolibarr95** [comment]
Hi all!
#8171 #6198 #2868
Welcome @ChangePlaces (sincerely)! If you have ideas/suggestion just make a PR.
Dolibarr have been created by french users, so of course there is some french. But as you see we  try to speek english (google translate helps lol). There are dolibarr forum in other langages.

**Comment by ChangePlaces** [comment]
@dolibarr95 hi, most of the videos I've been watching on youtube have been in french and I've relied on the broken auto-translated subtitles youtube provides! It's do-able, but is still a barrier to learning and importantly understanding dolibarr. I see the drive to get a course created for dolibarr which is a great idea, but will this be a good idea as it'll get more critical eyes on the codebase which could point out the same things and drive people away from dolibarr?

@mwelters But here's the problem - why would anyone want to fork the code, when it's going to need completely re-writing and re-structuring from the start? It would need too much work for it to be a fork! 

I would love to make a PR, but everything would need changing - I don't have enough dolibarr system knowledge to even begin and as I'm a consumer on the search for an ERP which will need lots of additional code changes I would not feel confident using dolibarr as a long-term strategy.

It works for some people who want to make a change here, a change there, but that's only a small part of an audience. Business processes are unique to every company and it'd be nice if dolibarr could accommodate this.

**Comment by torvista** [comment]
I understand that investing time building your custom code on top of a codebase that you are not happy with, seems a bad idea.
But one has to look in detail at the project as is (i.e. the develop version, not the released version) and evaluate how much it is really lacking from what you actually **need**, and balance that against spending much more time on a better-designed ERP that needs a lot more work to get to the same point.
After much research several times over the last five years, I came to the conclusion that Tryton was the best, but I don't have the time or inclination to learn  and modify it, when I already am competent with php/mysql from working with my online shop.
One has to factor in the time and effort required to get to the end-point of production-ready code.

Dolibarr does not need a course to learn how to use it. If it did, the GUI needs to be fixed to solve that.
Every time I have found something poorly explained, I have updated the language file...to aid my employees, and updated the wiki where appropriate.

**Comment by PMickael** [comment]
I'm totally agree with @ChangePlaces , I also open a issue a few weeks ago about commits, the first things to begin with, it to make a real commit usage, it's a not CTRL+S shortcut, you need to think of what you do, and more-ever fix things over the time, I update some part of the code for my own needs, and new updates introduce new code problem/syntax, like as simple as choice between space and tabs which is actually a random pick depends of the weather of the day 😄 

And I think there is a big gap between users versions, I tried to stay as much as possible to the latest version, but it's really complicated because I don't have the time to look at all this random commits to see if I will not break like VAT module which was not working properly in ~6.0 => ~7.?.? and fixed now (I think, I don't test all the case)

All the problems point to commit usage, it's not possible to make a code review if things are not correctly commit.

**Comment by frederic34** [comment]
the rule is for week pair
```
if () {
}
```
and for week impair
```
if ()
{
}
```
and sometimes
`if () {}`
and so on... the way to strict PSR2....

**Comment by Thunderstr1k3** [comment]
Cheerio folks,
I just stumbled across this thread. I totally agree with @ChangePlaces. Dolibarr has a great usability on its surface, but as a matter of fact, it lacks features - just like any other ERP System. As a software developer I tend to write necessary lines, but french code is a show stopper. Though I had french in school for almost a decade it is impossible to read the code fluently as one has to switch between languages every few seconds. Seriously, being french is nice and all but no excuse for a messed-up code.

Since I really like how dolibarr is designed, I would volunteer for assistance. @ChangePlaces Shall we? Anyone else?

**Comment by nielsdrost7** [comment]
I'm currently helping to get an invoicing application into a new version (PHP of course), but I could help with this application as well, maybe even intregrate invoicing into Dolibarr.
When everyone agrees on development environment, coding standards (phpcs), I'll help where I can, just ping me (set up a slack channel, would help)

**Comment by cfoellmann** [comment]
I am also evaluating ERPs at the moment and on the surface dolibarr looks very promising to me but I definitely need to add some extensions. A bad codebase will cost me more time and my firm more money.
But if the project leads are on board with evolution of the codebase like: consistent coding standards and replacement of French with English "code"; I might be able to sell the project to my bosses

 
@nielsdrost7 
👍 coding standards (phpcs)
👍 slack (or sth similar)

**Comment by nielsdrost7** [comment]
So I've created an organisation : dolimodern. i've committed 4 repo's:
- [docker-dolibarr](https://github.com/dolimodern/docker-dolibarr) (and [docker-doli](https://github.com/dolimodern/docker-doli) which is a fork of someone's docker-setup for dolibarr)
- [original dolibarr](https://github.com/dolimodern/original-dolibarr)
- [modern dolibarr](https://github.com/dolimodern/modern-dolibarr) built with laravel and split into modules. View files are set up, controllers need to point to the view files.
If you guys want to build things in Symfony, you're more than welcome to grab the view files (blade templates). They're basically just html, since this is the very beginning of transferring Dolibarr to a more modern setup

**Comment by UnderDogg** [comment]
@ChangePlaces @cfoellmann on my local PC i've converted the database into something workable. I've generated models, views, controllers, migrations, seeders (using factories)
I'm willing to convert it onto a laravel project if you guys really want to.

I can't do things and then find out that you guys want to make the new version with Symfony.
I 'very added you both to the organisation, so you can edit all those repositories and make commits to them
@Thunderstr1k3 I added you as well (invited)

**Comment by BebZ** [comment]
Hello, just a "noob" question : 
What makes you think Dolibarr code or fuctionnalities are better than other ones like TinyErp, Axelor, or else... Did you already try to read other code like Axelor's one ? 
It seems on other platforms they've put as much as possible functionnalities, but they can make only minor upgrades because they are stuck with coding standards/platforms.
Dolibarr seems more free from platforms/ strict coding standards. Isn't it ?

**Comment by UnderDogg** [comment]
Tinyerp and Axelor are made in different programming languages, so it's more difficult to make changes.
There is almost no way to compare these 3 products.
Coding standards: when Dolibarr was invented coding standards didn't exist or barely existed. That's why the new effort needs to start with the information we have (coding standards is one of them)

**Comment by PMickael** [comment]
@UnderDogg a new codebase from scratch doesn't sound like a great idea, because I don't think that we get enough mind ressources on this.
Maybe a cleanup of what exists will be a start we do not actually need to use a framework to do that, we just need to get access to the codebase but it's seem's to be complicated...

If you want to start a new CRM ok, but it will not be dolibarr anymore

**Comment by jtraulle** [comment]
On the same page with @PMickael ... A hard fork and complete rewrite from scratch seems absolutely unreasonable and will fail like the few forks before that.

**Comment by cfoellmann** [comment]
I also don’t see a future for a rewrite.
It costs to much manpower. Starting something new is great but something so complex will take a long time until it would be usable for the users of an ERP system.

I am not sure what route we are going to take. Even if we pick up dolibarr.
Main plus for me was php and MySQL and with a commitment to cleanup and „translations“ it might still be a viable option.

**Comment by dolibarr95** [comment]
![rewrite](https://i1.wp.com/tech.ccmbg.com/blog/wp-content/uploads/2014/05/BoCOJdJIQAEzod7.jpg)

**Comment by hregis** [comment]
For my part, I consider Dolibarr as a framework, it's up to us to improve it. 
Make a break in the engine will generate a lot of problems! 
See Prestashop who has made a break in its mode of operation, it involves fees to the user! 
Dolibarr is an ERP/CRM! A cut would be fatal!
We made a fork, which is no longer a fork, because completely rewritten in NodeJS + MongoDB
https://www.tomanage.fr/

**Comment by dolibarr95** [comment]
@nielsdrost7 
https://www.dolibarr.fr/forum/15-retours-dexperiences-utilisateursintegrateurs/64477-dolimodern

**Comment by UnderDogg** [comment]
Guys it's ok, it was just an example. I took all of Dolibarr's HTML files and did put them in a template, so you can generate templates from the Dolibare source code that you want to improve.
I just read the opening post and when you look at the repositories you see the examples.
- grab the docker environment and run it in the usual way.
You guys probably want to improve the current Dolibarr sourcecode and that is perfectly fine.
- unfortunately all information is echo'ed right from the source code (the 2005 way, the title of this topic). When you want to do it in a modern way, you want to make templates (views). If you look in dolimodern you will see all those views.

I wish you all the best of luck with improving the current sourcecode

**Comment by hregis** [comment]
it's possible to reinvent the wheel, 
but why reinvent the wheel when we could improve it? ;-)

**Comment by cfoellmann** [comment]
I really like @UnderDogg s drive.
If we can iterate and keep everything backwards compatible it could work.
Improve „everything” under the hood and keep it running for everyone invested in the project.
We have to keep in mind that using an erp System is a long term investment and you need to give the users (enterprise) the peace of mind to be able to rely on it for years.

If we could move everything to php 7.2 (+5.6) + horizontal scalability via docker it should be a good idea.

The templating system (MVP concept) while keeping extensions working “for now” in parallel should be a good move!

I hope this discussion keeps going. And keep it civil people!!

**Comment by hregis** [comment]
the 9+ version of Dolibarr does not have any problems with php 7.3! 
The VMC can be broken down into many concepts!
The rules are there to follow a road, but is it the right road? ;-)

**Comment by cfoellmann** [comment]
I looked at the code and it is a horror. Especially when you do not understand ANY French!
Rewriting something like the templating would require touching a large portion of the code base. Ergo many of the French “elements” can be replaced.
An English code base would allow more people to contribute

**Comment by jtraulle** [comment]
Maybe it is possible to progressively transition from French to English in existing codebase (without the need of hard fork). If this is done one module at a time with the appropriate SQL migrations, it should be doable ... What do you think ? On the other hand some core modules are tightly coupled and should be refactored together ... One other major problem that I can think of is about preserving compatibility with the existing external modules ecosystem provided by various developers ...

**Comment by hregis** [comment]
@cfoellmann 
I develop on Dolibarr since 2005, it is a French project at the base, so inevitably there is French!
We modify little by little the code to put English from everywhere ... without integrating bugs !! ;-)

**Comment by torvista** [comment]
I wish Eldy would comment on this...as there is a lot of energy, time, effort and experience being wasted just in this thread as it is in a vacuum.
I think my first posts are still relevant.

**Comment by ChangePlaces** [comment]
It's interesting to see this thread come alive again, and I think it's provoking some interesting discussion.

@torvista an ERP needs to be easily modifiable to fit into business workflows. One size does not and can not fit all. In addition, your other point that other people don't contribute to dolibarr is probably because the code is a mess, not understandable (both because of French, and no real order / sequence to the code).

@UnderDogg I don't think a hard fork is appropriate. There are 777 issues. If you restart, you're going to be fighting those issues and creating many more. Also, symfony would be a smarter choice. Laravel is too opinionated for its own good.

@dolibarr95 : I'm sad to see the bitchiness and catty attitude from you towards those trying to help an aging project into the 21st century. It could breathe life into dolibarr and open up many new doors. e.g. having an all-English code base would allow many other people to wittle all those issues down to a smaller number. Everyone would prosper here.

Phrases like, "it is what it is" and "it works for me", and "I don't want to change it much" are not good enough and are the bain of advancing past the fear of the unknown. Imagine having a codebase with plugins... and how other developers could develop their own plugins to add new features and even opensource them for dolibarr to use.

Well-planned, thought out changes to the code base would make no difference to the end user what-so-ever and could be planned over a period of time with people helping out where they can. Database changes could be handled effortlessly with migrations, and doctrine would make this a breeze. Some of the earlier commits would be horrid as a simple change would have to encompass so many files, but over time this would improve.

It needs support from core developers who have an intense knowledge of the system and a community wanting to help. It seems like 50% of this is met so far!

**Comment by hregis** [comment]
@eldy 
Eldy wants to keep maximum autonomy from external libraries, and maintain compatibility with devices for the blind peoples! Dolibarr is already a Framework! It must be changed little by little! No break in the code!

**Comment by torvista** [comment]
> other people don't contribute to dolibarr is probably because the code is a mess, not understandable (both because of French, and no real order / sequence to the code).

Well that is your opinion, but my opinion is that most people who start using Dolibarr are small business people who are not aware that the code is not built on a 2019 framework.

When I started preparing it for my business and started digging into it, yes there is a mixture of french and english, but that that is only to be expected and it makes no sense just to rename things unless that is part of a code improvement: if it is not broken, don't fix it. There are plenty of things that DO need fixing and they should have priority.

As I said before Dolibarr is a large project with a long history and lot of legacy code. I have no doubt the developers would love to change it all overnight but that would destroy the  user base who do not have the skills to manage a big bang/breaking upgrade.
But, a detailed roadmap from the developers to detail how they ARE going to move the project codebase forward is, I think, solely needed, to better focus a discussion such as this.

Where real developers as yourselves are showing interest, the project leaders should be jumping at the chance to include you and guide your efforts to where they think they would be best employed.
Their lack of comments is not encouraging....

**Comment by hregis** [comment]
@torvista 
Think again! it's encouraging ! 
I had doubts a few years ago and a fork was born ... 
but the evolution of the Dolibarr code is such that it makes you want to contribute!
Dolibarr moves fast and well!
it must be considered as a framework!
A base that allows you to create what you want!
an external module can use the technology we wants!

**Comment by dolibarr95** [comment]
>  bitchiness and catty attitude from you


@ChangePlaces the picture is a joke sorry if i hurt somebody here..and I just invit nielsdrost7 to comment in the forum in a specific topic about his work.
;)

**Comment by dolibarr95** [comment]
@ChangePlaces about videos...
As I said here : https://www.dolibarr.fr/forum/12-howto--aide/62681-wiki-dolibarr-org
My objectiv is a perfect wiki (full documentations, screenshots...). I've got my opinion about videos, mooc, and other platforms...but this is not the right place to talk about it i think.
:)

**Comment by PMickael** [comment]
Core dev team right now :
![don't care](https://media.giphy.com/media/PTl6ImjH2qBrO/giphy.gif)

**Comment by FHenry** [comment]
@UnderDogg I give a look at you code, nicly done, install it, but I always get 404 page not found of laravel framework, maybe welcome or home route lead to nowhere ?

**Comment by dolibarr95** [comment]
:) It could be nice if can @nielsdrost7 present/talk about his work :
https://www.dolibarr.fr/forum/15-retours-dexperiences-utilisateursintegrateurs/64477-dolimodern
or create a topic in the englis forum

**Comment by UnderDogg** [comment]
@FHenry interesting, can you go to /ownercp? Or /projects? I'll take a look as soon as I can
@dolibarr95 on my way, I didn't have an account and my French isn't that great, so I need google translate
I did some replies on the forum, I hope they understand

**Comment by ChangePlaces** [comment]
I can't create any issues in doli modern, so here goes as a potential roadmap overview:

0/ It's been a while since I've looked at the code base (been busy creating my own system) but if unit tests do not exist, they need to be written. No one like writing them, but they always uncover things that are broken)

1/ A familiarity with the current issues should be noted. You may come across solutions as you touch the code base.

2/ the current codebase should be converted to english as a priority. this can be done with online translation if necessary, but I'm sure there are enough native speakers here to help. At this stage, it should not translate table and field names, but should include where necessary: comments, filenames, variable names, class names. Comments should be done first. The easy part. Next, a good IDE will replace the names across the whole project (though they are not perfect, so a commit should be done after each variable / class  change just in case). Care should be taken to avoid changing things that are not related to the current change. This is where core committer support would help, but, beggars / choosers etc. This is not an easy step as care and attention has to be paid to the renaming of the variables / classes. When all tests pass, all's good.

From here things will get a little fuzzy. It's possible to merge twig templating into the system with the current code base (Symfony doesn't enforce the MVC paradigm) and get the same output. Templates would have to be created and move the necessary logic in to either twig language, or in to controllers and then pass the data to the template (depending on how complex the data is). Sub templates for things like tables and other view can be crafted and pulled in by other templates when needed. This would simplify templating by 100 times!

A single index page will use a "router" to know where to pass the request to. This will be passed to a controller which will pull in the necessary data then calling the templates. The codebase from above should be able to be wrangled from many different files that call the above templates into several single-concept controllers. It's handy to have one set of controllers for the user views, and one set for the API.

The API is not to be left behind, the controllers in symfony can easily output json data (simply `return $this->json($data)`) and should return the same data as currently. This will need to be 'bodged' in places, as changing the api (if necessary) to english names will be reserved for later as it'll be a breaking change.

Then, models can be created using doctrine and then integrated into the above View/Controller system. This allows annotations to create the meta data of the database. Here, table names / field names can remain the same (it's invisible to the common user). As changing field names can be a breaking change (e.g. 3rd party code which interface directly with the database for private stats etc would break). It would have to be reserved for a major version change.

---

Each of those steps can be broken up into 100 different steps, but for a general overview it's a start.

Project structure is important, e.g. in the current code, to install plugins you're overwriting code unknowing what it's replacing it with. Huge security hazard. By putting modules inside a 'module' directory, will prevent this. Although not a priority at this stage, it'd be possible to separate the modules into separate projects so they can be installed as necessary by the user.

I remember reading about the background of dolibarr and how its developer/s received support from the french government. I'd assume if you're not being told to make a change, because the boss is happy, then why bother wasting energy changing it?

**Comment by altairis-tof** [comment]
Am i the only one to think that an ERP is made for users and not for developers ???
Dolibarr is very popular and i dont understand why some guys want to break it into pieces... 
if you want to play on your own, just start a new project but please dont waste our time !
if you want to improve dolibarr and follow the lead, you are welcome.

**Comment by ChangePlaces** [comment]
> please dont waste our time 

who asked you to "waste your time"? Improving the codebase will mean more people able to contribute which equals more features, better code, fewer issues with the code and hopefully fewer bad attitudes towards change. Essentially the changes would be invisible to 'you' the end user, but as you're on github you're more than a 'user'.

> if you want to improve dolibar...

That's the whole point...

>...and follow the lead, you are welcome.

From a development point of view, there is 'no lead'. The code is an awful mess that only some people can contribute to, and oh, I'm repeating myself again

**Comment by altairis-tof** [comment]
ok for improve; not ok to throw it all away. and yes i'm a developer and i know that what i develop is for end users not for developer pleasure. and last more, french is one of the most spoken language around the world ;-)

**Comment by cfoellmann** [comment]
I think we should return to a more productive discussion!!

@altatof About the reach of French for _users_ (people) you are right but if you were to measure the language skills among developers you will get a different picture.
> i know that what i develop is for end users not for developer pleasure
Your result is for the end user but in an open project you do have to consider how your work is impacting your fellow devs/contributors

**Comment by ChangePlaces** [comment]
I'v'e just re-read down this thread from the start and noticed a few things:

- @eldy removed the "bug" from the title of this issue and sees it as "code enhancement", 
- there are a series of vocal 'developers' who are aghast at losing the French contained in this code as if it's some sort challenge to national pride, 
- there are issues going back to (at least) 2015 about re-organising this code base which have either been closed, or are still open, either way no action has been taken.
- no support from core team at all.

The last point is a concern. If they don't advocate clean code, then I dread to fear what undetected issues are happening. 777 issues is a lot for any project and the core team / person should be jumping on the chance of improving the codebase.  But they are not.

As it stands, it's unmaintainable, impossible to merge into any business practices, impossible to read (even for french nations), and I would rather exert my energies to other, more productive situations. I've seen this before with Drupal (such a complex code-base no one could understand it except core dev team) which now is mostly symfony modules, and prestashop (core dev is offensive to people who point out issues with the code) which has serious security issues, and more.

As I wrote elsewhere, without core team support, any large changes to the codebase will be impossible and a waste of time as it'll be fighting them as they'll want to modify their code, not the improved code.

It's a shame so much energy has been in this thread.

**Comment by torvista** [comment]
> no support from core team at all.

THIS is the FIRST issue that needs to be resolved, anything else is irrelevant/wasted effort until there is some conversation/signs of life from them.

**Comment by cfoellmann** [comment]
@ChangePlaces thanks for the summary.
I am totally with you on your results.

In my quest to find an open ERP system written in PHP Dolibarr was more or less my last hope and on the outside it looks promising but due to the issues @ChangePlaces listed above it is not looking like I can go with the project.
Since I come from the WordPress world I know a lot about doing the "not-standard" way or hard to convince maintainers. But WordPress is a project with huge legacy support but high quality releases. Despite what modern projects say about it it is dependable and promises long-term stability.
That is what I was hoping to find in Dolibarr, too.

**Comment by cfoellmann** [comment]
> 777 issues is a lot for any project

The 147 open pull request show an even more grim picture. PRs that just waste away will discourage contributions and hurt an OPEN project. Open source is a choice that need to be actively engaged!!

**Comment by cfoellmann** [comment]
Me again 😉 

I know that the French are very protective of their language BUT I dont see why changing the codebase to English exclusively is sooooo bad!?
In the current state a French-only speaker does NOT understand the code because comments and code are partially English (whatever the ratio) and an English-only speaker does NOT understand the code either because of the Frensh portions.

So my question: What good does it do to reduce the group of devs able to contribute?

**Comment by altairis-tof** [comment]
it is not to be protective of a language or another; personnaly i dont care if the invoice table is called "invoice", "facture" or anything else; the fact is that end users dont care about it and want to have an ERP that suits their needs, whatever language or framework is behind; we have a good ERP that works with lot of users, and lot of coders at this time. So if you want to make it evolve, just do it by making PRs ;-)
By the way, this kind of thread (troll ?) comes back every year or so, but i did not see many real contributions apart from discussions from those who initiate them.

**Comment by torvista** [comment]
I think the comments in french are irrelevant...I have seen they DO get changed when the file is updated, but really there are more important issues than that. You have to accept the history, don't use it as evidence the project is bad because there are non-english things in it, that is very arrogant.

I think the code/variables named in french is a minor irritation for non-french speakers...but changing them globally is major disruption...it can be done, but for what real improvement?
It does not mean the code is bad.

BUT:
777 issues...unresolved, although the great majority are probably fixed...but who knows, yes it shows a lack of interest, professionalism AND a lack of pride!

147 PRs left for dead...yes it shows a lack of interest and yes, discourages contributions.

No comments here from the development team...in fact IS there a team or only Eldy?...it shows a lack of interest.

Sad, it makes me drag my feet about developing this.
Like many others. I am hoping something else appears...but you know it won't, not overnight, to compete with the features Dolibar has: that takes years of development.

**Comment by cfoellmann** [comment]
for me the "bad" code (if it is or not) and legacy is not the major part giving me pause here.
I love refactoring, modernization and code review but I learned my lesson in the past. Nothing is worse than sending PRs and getting back nothing. It will bug you every time you look at your open PRs list on Github.

If there was official movement here I would be more than happy to get my bosses to sign of on starting to evaluate Dolibarr further.

**Comment by ChangePlaces** [comment]
I'm going to go ahead and unsubscribe from this thread. Bon voyage!

**Comment by eldy** [comment]
Dolibarr has (when i write this post) 138 open PR on 8291 closed. On the 138 PR still open, nearly 95% have an answer with a comment or a github tag, so Dolibarr project has a rate of PR without answers of around 10 PR on around 8000 answered and validates. Only 1.25 % of not answered PR !
This is probably a the best rate among any project of the same size (more than half a million of line of code if we count only the core code excluding any external libraries). 
Of course, we can always do better. But with a such good answer rate, it is difficult for a project that is 100% managed by a huge internationnal community to do better (most project have a rate higher than 5% of unanswered PR).
So, any idea are welcome. New features of Github, more strict development rules, automation tools added frequently helps us to make this rate each time better, but the 0% is probably something we will never reach.

10 years go, the project was 100% french.  Contributions to migrate to english are always welcome (English is official language for any new contributions, this was decided 10 years ago). We are also on the good way for that. Currenlty, 90% of code comments in v10 and 70% of fields in database are already in english. And after each version this rates are better. Currently most developers that does not speak french (more than 50% of the community) say this is not a priority for them, but we plan to reach the 100% despite that.

**Comment by BebZ** [comment]
Dear all, I would follow Eldy... And his pace! 
if you would be here since 8 years, I bet you would respect his way to go without all these standards every lazy loves... At last it's not the framework who corrects... It's mainly the Guru, the Yoda... Or the one who deeply knows the code! 
As an user and/or a company that didn't find everything wanted in llx_tables... but that has looked already eeeevrywhere for better... I didn't find aaaanything better than "self controlled code" ...

2cents for those who want to "change p(l) ace", ur welcom !

**Comment by PMickael** [comment]
@eldy any idea, then like everyone here ask for the same things, get team access on this repository like any other open source project to be able to make big upgrade on the codebase, not by PR.

With this access we can discuss in Github team discussions tab about what's next and let people that want to contribute be able to.

**Comment by cfoellmann** [comment]
@eldy thanks for joining this discussion. I hope you are not offended by "all this". Even if the tone might be not nice in some places it shows that people (want to) care.

I am with @PMickael in some way. Extending the maintainers list might give the project some momentum.
Maybe a modernization like @nielsdrost7 can be evaluated in the dolibarr organization (dockerization) and in branches in the dolibarr repo.

Is there a chat for dolibarr? Like Slack, gitter, or sth similar?

I have to say that @eldy's comment on replies on PRs and issues showed me that my thinking was flawed. Thanks for getting into this issue @eldy

**Comment by torvista** [comment]
What about a plan of action for the ~800 unclosed issues? 
Reducing that to less-scary total would go a long way to calming the nerves of people thinking about using and developing Dolibarr.

**Comment by jonbk** [comment]
I would like to add my point of view on this.

The issue is two years old now, and I've the impression that an agreement is impossible between legacy developers on Dolibarr and guys who want to rewrite it.

First of all, I would like to start with: Dolibarr is **GREAT**, and I would like to thank all of you who contributes to it since it started in 2005. Especially @eldy and @hregis who are still here, almost 15 years later. You guys did a tremendous work and can be proud of this product used by many people (including mine for several customers). Really, **GREAT JOB!** 👏 

Now here is my developer experience on the projet. Each time I develop a feature for a customer, I try to stick the most closely to the core for maintenance purpose. And each time, I found it difficult to retrieve the files where the targeted function is, what dependencies I need, how to include it (`include()` hell). It's difficult to handle the HTTP flow of the request: many and unconsistency data in `$__SESSION`, `getpost()` function which is awful in my opinion (you want a GET or a POST parameter, not a function who choose for you where the parameter is.). Security is a mess: constantly adding checks in each file. Add to this the many ways some pieces of codes are written for doing the same thing in different files, modules,... Creating third party module is HARD, despite the module generator, there are many files with verbose code just for create a manifest of the module. In addition, it requires to block an ID of the module on the wiki page, to be sure that two modules are not in conflict. This is clearly not a convenient and secure way to do that...

I could list many more examples, but that is not the goal.

All these observations lead me to the following conclusion. Dolibarr is a great and useful tool, but deserve much better today, and could be an even better tool if we engage a refactoring project around it. I personally think that such a tool which contain sensible data, should be more solid and with well tested functionalities, in order to be opened to many more contributors and increase sensibly the notoriety of the project. This could be a great business opportunity for many people to propose a very solid framework to public.

To achieve this vision. A major refactoring is, in my opinion, requiered. We should establish standard/guideline chart who every developer should respect. The contribution of a framework like Symfony (which is the perfect candidate for such a projet in my opinion) could be incredibly beneficial for such an initiative (HTTP flow, security, MVC structure, Service injection, Twig and Forms, Debug Profiler,...).

The danger for me is to see Dolibarr slowly die because of lack of contributions (if @eldy leave the projet, we're dead). Legacy contributors should accept that because of the success of their tool (which is **NICE!**) , they have to federate a biggest and various community to ensure the sustainability and success of Dolibarr in the future. And start coding on a solid codebase is a requirement of this.

Maybe we should create a direct messaging account on a tool like Slack, for starting discuss about how we could engage such a projet around the actual Dolibarr, hand in hand, together.

I hope I didn't offend anyone. I just want the best for Dolibarr, and this is my personal opinion, who appears to be the opinion of more and more developers here.

Dolibarr is great, and we should pamper it!

Thx.

**Comment by jtraulle** [comment]
Thanks @jonbk for this genuine and balanced review :+1: 

I share the same thoughts. 

I think even some "Legacy contributors" share your point of view (I hope so) but we need to be realistic about the feasibility of such a huge project. This is really, really hard because of the existing ecosystem (external modules that rely on the existing architecture, really tight integration between existing core modules etc.).

Some things could be however easyly (sort of) achived in my opinion like

- [ ]  using namespaces and autoloading instead of includes, require, require_once
- [ ]  using composer for managing and locking dependencies versions

BUT, I want to say that this is something completely feasible. Others opensource projects has already started this long and difficult journey. I think particulary about PrestaShop that is in the process of migrating to Symfony components / Doctrine entites and Twig template for the backend (this is the good thing about Symfony, it is possible to use only some components and engage in a refactoring project one step at a time). So this can be achieved step by step, one module after another.

**Comment by jonbk** [comment]
@jtraulle I would be glad to discuss about this, but in a more convenient way like Slack. Is there any possibility to open a Slack account for Dolibarr contributors ? 

I do not know who tag for this request ? Who is ine charge of managing community tools ? @eldy  ?

**Comment by jtraulle** [comment]
There is already a Discord server here https://discord.gg/gCM77Yy

There is also a slack that has been created on dolibarr.slack.com (see this related issue https://github.com/Dolibarr/dolibarr/issues/12028) but I do not have an account on here and the invite link is expired.

If we go for Slack, we should pass the [Slack for Non Profit organizations](https://slack.com/help/articles/204368833-Slack-for-Nonprofits) to lift limitations of the free version (only available history for the most recent 10 000 messages). It worth noting that even when validated, workspace is limited to 250 free members).

Another option is [Keybase](https://keybase.io/) that is free for teams of all sizes.

Not sure who is in charge of managing community tools but, yes, @eldy is the project lead/maintainer

**Comment by jonbk** [comment]
Disord is OK for me, I join 👍

**Comment by eldy** [comment]
Communication tools are currently managed by the Dolibarr association that is not affiliated to development core team (even if some people are part of two)

**Comment by geri777** [comment]
I would really love to see this project more sucessfull! It is the only relevant ERP system using PHP. For me the most important issue is the bad documentation. My suggestion would be: create Youtube tutorials (in English) starting with:
- Installation
- Basic Setup for an imaginary business like a cookie store. 
  - Product Setup
  - Bank Accounts / Financial Setup
  - Customers and Vendors
  - Stock
  - POS to sell a product

The current videos on Youtube are a) outdated and b) in French and c) don't cover most parts a user needs to know to start with Dolibarr.

**Comment by dolibarr95** [comment]
Hello
Join us in the forum and share your point of vue and ideas :
https://www.dolibarr.fr/forum/t/tuto-video/23095

**Comment by dpriskorn** [comment]
> For my part, I consider Dolibarr as a framework, it's up to us to improve it.
> Make a break in the engine will generate a lot of problems!
> See Prestashop who has made a break in its mode of operation, it involves fees to the user!
> Dolibarr is an ERP/CRM! A cut would be fatal!
> We made a fork, which is no longer a fork, because completely rewritten in NodeJS + MongoDB
> https://www.tomanage.fr/

This link is now dead, which proves the point of difficulty making a hard fork.

**Comment by dpriskorn** [comment]
> What about a plan of action for the ~800 unclosed issues?
> Reducing that to less-scary total would go a long way to calming the nerves of people thinking about using and developing Dolibarr.

I took a stab at the issues today. We are now down to ~540, but a lot of those are stale and discussion and feature requests. 
Here is some statistics on the issues:
261 bugs open (only ~76 have been [opened after v12 was released](https://github.com/Dolibarr/dolibarr/issues?q=is%3Aissue+is%3Aopen+label%3Abug+created%3A%3E2020-06-21+) and my estimation is that only ~20 of the earlier bugs are lingering bugs also present in v12)
5 bugs need more information (only 2 relates to v12)
The rest are feature requests and discussions issues not labeled correctly that have not been closed automatically yet.
This is really not that bad for a project of this size and complexity IMO. 
Horray for Dolibarr!

**Comment by dpriskorn** [comment]
I have now labeled all remaining issues. No issues without labels should be left. Lets squash some bugs!

**Comment by Foorack** [comment]
@jtraulle
> There is already a Discord server here https://discord.gg/gCM77Yy

Yes, but that is entirely in French. Contributing to this project is very difficult when not speaking that specific language, and it discourages a lot of developers worldwide from helping.

**Comment by cfoellmann** [comment]
@Foorack I am with you.
The slack fell through. But I would use the discord if there were some "management" to provide itnernational channels. (I am not sure how those are called in Discord)

**Comment by Foorack** [comment]
@cfoellmann 
> But I would use the discord if there were some "management" to provide itnernational channels. (I am not sure how those are called in Discord)

Yes, that is possible. You can have language-specific channels, commonly by appending "-en" or "-fr" after the channel name. It is also possible to create Categories of channels e.g. "English discussion" and "Discussion en français". This needs to be made by moderator with sufficient permissions or server owner.

**Comment by cfoellmann** [comment]
@Foorack 
The problem is not the technical aspect. It is the engagement of non-French speakers. Just like you said it discourages devs from contributing.
These channels need to advertised on the official homepage and forums so "normal" visitors know these are available

**Comment by cfoellmann** [comment]
Maybe we can close this and concentrate on the big number of open issues.
I propose some kind of chat meeting to go through "all" open issues and close or assign to someone. To move on from this discussion

**Comment by bitsnaps** [comment]
I've see that Yii2 is very stable framework, as it does provide LTS versions, with support old PHP along with new one, even Yii1 is still supported, Yii3 is coming with more modularities, clean base code well tested, PSR compliant, although other frameworks are more into hype...

**Comment by cfoellmann** [comment]
I think moving the core to a framework is like starting from scratch. This seems to be a non starter.

The project is moving faster than I thought in since we started using it.
While I love getting new features I thing the project should focus at least 2 major releases (1 year) on improving what is already implemented and modernize some of the core classes (most of which are not even classes yet) to make everything more consistent.

I "everything" would depend on internal APIs for eg FORMS it would be possible (not needed maybe) to move these over to a framework. But creating an internal set of core APIs using classes for different purposes throughout the codebase would be modern enough and great for the future.

**Comment by torvista** [comment]
I think the project is continually running before it can walk, introducing semi-tested/coded features while there are a vast number of unclosed issues. You can't do that with a business-critical application.

Personally I would stop adding features to the core until all bugs and issues with current features are resolved.
In this interim
- new features should be implemented by those that need them, as add-on modules.
- a framework migration is a very long term project but it needs to start at some point, running in parallel/interfacing with the existing code.

**Comment by fappels** [comment]
Think future Dolibarr is 'Headless Dolibarr' like 'Headless Wordpress'. Mass extend REST API and make new front-end that runs on client.

**Comment by cfoellmann** [comment]
I am with @torvista mostly. That is why I suggested a feature-freeze to concentrate on bugs and improvements. Also many features are not complete or even usable in real life. These issues need attention and resolution.

I do not know how but who can help on getting together a bug-squeeze meeting regularly? We need bi-weekly (or less?) meeting free for all where "all" open tickets/issues are looked at.
That way, issues can be closed, properly tagged and assigned to maintainers.

About going for features as add-ons: That is what I think is a proper way BUT: while devs are going to do this, we need to make the core more flexible and extendable where the addon devs need it. The core code is so static and a mix of frontend/backend code that even simple features can NOT be implemented by addons.
SO:
1. Yes, all new features only as addons!
2. "All" feature addons will show where in the core we need more flexible and new code. -> interfaces and APIs must be implemented for addon development as fast as possible when requested!

**Comment by Dom911** [comment]
Hi,

I am not a developper. My past carrier is about Managing Software.

As already written, the project DOLIBARR is great. I am using it for projects and accounting and I am even thinking of providing as a Saas (but with Kubernetes and Docker, not with the Saas project on which I could write during hours. It is a good intention but ... ).

DOLIBARR now suffers from a good architecture. In 2005 there were no framework, and it is as it is now. 
We are in 2021 and there are good frameworks that would simplify the life of everyone (just check the kind of question on the forums for development and you will agree with me with no hesitation). 

For example: I just have tried to create a new module. Even if the builder creates "the basics" it takes sometimes days just to understand how to add a simple thing like a menu (which by the way does not work at first in the auto generated module, neither the loading of my translated sentences). (I repeat: I am not a developper but I have very good basics)

Software updating is part of the Continuous Improvement.
Moving the code completely to Laravel or Symphony would require a huge manpower, would take months of development, would create tons of bugs with the risk of losing developers and users. So, NO GO.

Not moving is not an option neither. The more Dolibarr grows up, the more it will be difficult to change it in the future and the more it will be hard for new comers to understand it. 
Someone has written, if Eldy is no longer part of the developers then Dolibarr is dead !! This cannot be. This is the **HIGHEST RISK** that a software can have. The only mitigation is hence to move to a framework understood and mastered by more than one person (moving the knowhow to one or 2 people is not an option).
So the question is not if we move or not, but where do we start at first.

Moving requires a GOOD DOCUMENTATION and a PLANNING.
Developers need to understand the current state, the new classes to be used and how, and the deprecated methods that will come in the next future. 

For those who thing that all the methods should be available for ever, the answer is NO. Updating your code is normal.
If you want something that does not change, then change job. 

In my opinion, the best way to start is with the generation of new modules that could - step by step - slowly introduce the Laravel framework (as Symphony seems dead... even if SymphonyRock has been created, Laravel seems taking over by far).
An App would be here the module. We could then create private bridges between the current architecture and the new one.
If only the 1st mechanism App / Router / Controller (with or without / View / Model) could work, that would be a good start.

Third party modules need to be updated accordingly and old modules need to be marked (like WordPress does for example). 

There are many tools that could be useful instead of creating the wheel each time.
Maybe we should also start to list them. Projects like :
- Blade for templating
- PHP Intrusion Detection System or HTML Purifier for security
- PHP Codeception_ for testing (PHPUnit could be used for test regression/assessment)
- PHP Mess Detector
- PHP Copy Paste Detector
- Sensiolabs (for security testing)
-  I do not know if there is a Continuous Integration currently. If not introducing Semaphore would be good.
-  ...

One important point are the environments : developper, tests level 2, [ ... , integration,] deployment.
I am using docker and was able to have a development environment ready in less than 10 minutes.
If I stop docker, I have choose to have the DB erased created again and I can start from a fresh environment in less than 2 minutes.
Having the different environments ready would be a real plus (development with debug, SQL sniffing, security checks and git available,...  tests with CI included (where do the reports go?), deployment for end users).

My last words : In order to success, start small, but continuing moving is the key.

**Comment by Dom911** [comment]
Thanks to @jtraulle the english channel has been started on Discord.com. 
I know, you must pass the sign up in French, but the channel wiil be in English.
https://discord.com/channels/529709484797591563/897856388548673567

I think that we will need project management tools also, but before to propose something I would like to know what exists sofar.

**Comment by rsanjoseo** [comment]
I have started to do a refactoring of Dolibarr, but the code is too outdated.
https://github.com/rsanjoseo/alixar
And I say it as an admirer of the project.

**Comment by cfoellmann** [comment]
@rsanjoseo when I started to get into Dolibarr I was somewhat shocked by the code base compared to WordPress which is not the most modern and clean application.
But over the last 2 years I realized that most of my input and work is appreciated by @eldy and you can help make it better by sending your PRs to the repo.

Your modernization efforts would make a bigger difference if you would invest in the progress of Dolibarr.

**Comment by hregis** [comment]
@rsanjoseo It is preferable to propose this type of modification little by little in order not to break the functioning of Dolibarr, which was the case for Prestashop 1.6 to 1.7 or Magento 1 to 2

**Comment by dpriskorn** [comment]
> I have started to do a refactoring of Dolibarr, but the code is too outdated. https://github.com/rsanjoseo/alixar And I say it as an admirer of the project.

I looked at the files in https://github.com/rsanjoseo/alixar/tree/main/htdocs/Core/Base and I must say that it seems like a very good start. I liked the quality level of the comments!
Even though my PHP fu is close to zero I understood almost everything. :)

**Comment by Dom911** [comment]
Hi,
I had many thoughts on how to modernize Dolibarr.

I started to integrate Laravel (free well structure framework) into Dolibarr but, even with an approach of (Laravel inside Dolibarr and no the reverse) I ended up with major modifications of the basis in Laravel.

Dolibarr inside Laravel would give far too much work to rewrite at first.

The way I have found : I write my APIs in Laravel and I use (consume) them in Dolibarr. So I only use the Dolibarr framework for the display and some other stuffs, but the main functionalities (including data access, logic) of my App is in Laravel structured as I wanted.
This needs some configurations in Apache server (Virtualhost) + .hosts , but it works fine.

If you want to modernize Dolibarr, I think that the way I do would be the best. This would mean that you first write the API code in Laravel for the main functionalities, keeping for the moment the GUI.
You could progressively replace the code as you go with the API calls
- Authentication (a hard part)
- then the core (settings, companies, products, ...)

On the other hand, for the GUI, if you have a set of APIs, it will then be possible to use whatever the developer knows (Flutter, PHP, Java, ...).

**Comment by dolibarr95** [comment]
Hello 🙂
@Dom911  API is the future !!

**Comment by cfoellmann** [comment]
@rsanjoseo I also admire your investment and code but as @hregis says that big breaking changes do not make a project greater.
WordPress is soo big because of the backwards compatibility and iterative changes.
Your rewrite would need to be compatible with the "old" modules out there and I dont really see how that could work. Otherwise you kill off all extensions and all the knowledge of the current developers.

**Comment by Dom911** [comment]
@cfoellmann : I do not share completely this view. Yes some backward compatibility must exist. But take the exemple of PHP, they update the code and progressively remove some functionalities which have no sense anymore.
WordPress does exactly the same.

If the communication is well done, meaning that developers have time to adapt and understand what they have to change then the backward compatibility can be limited. Furthermore, this would also clean old add-ons never updated...

**Comment by cfoellmann** [comment]
@Dom911 I do not see it different from what you say.
I am not for everlasting support.
But php and WordPress support very old code! Most code written for early php 5.x is still compatible with 8.x. 
Like it was suggested removing some very old and outdated APIs would weed out bad and unmaintained extensions.
Despite the modern and very clean rewrite we can see here it will (please correct me if I am wrong) kill off all current extensions and devalue all knowledge the current community has.
Also an ERP System can not be handled as flexible as a Website on WordPress!
The update process of dolibarr has always been flawless for us which is the most important. When I compare it to other platforms where Update dangers prevent companies from using the most current version of a software.

Just some rambling...

**Comment by dpriskorn** [comment]
> Also an ERP System can not be handled as flexible as a Website on WordPress! The update process of dolibarr has always been flawless for us which is the most important. When I compare it to other platforms where Update dangers prevent companies from using the most current version of a software.

I totally agree to this. I run Dolibarr in production since 2,5 years for my own small business and it has worked flawlessly. Well I have run into a few small bugs and missing features, but nothing I have not been able to handle via forum, issues or programming my way around it.

I am using the API extensively and have programmed a few direct inserts/updates into the database when the API did not support what I needed.
The database structure is quite simple and I like that a lot (coming from Odoo 9.0 and 10.0 where everything seemed overcomplicated and they rewrote the whole thing)

**Comment by Dom911** [comment]
@cfoellmann & @dpriskorn : Maybe I need to explain better. 
Today we are facing a code which is getting old and we know that moving to a well structured framework (which came years after Dolibarr was written) is not that easy. 

So the idea is to use APIs and provide a way to use new tools for the GUI.
But first, let's concentrate on the APIs.
For sure there will be some modifications, but as you mentioned, we need to keep the current version working. 
It is a must for all those who have made some integrations. 

You were speaking of WordPress that is still having some WP 5.x code working : it depends on 2 factors:
- the number of developers we have (because we need workforce to maintain the APIs, create mechanisms so that they still work... the older, the more complicated)
- the number of users/third parties we will lose by API change (meaning that the developer will modify its code or abandon)
If you can mesure those 2 factors, then you will know where is the trade off.
If not... then you need to take a rule based on approximates. 

I guess that the number of Dolibarr developers is " slightly " different from WordPress, so if you want to keep 3 versions of APIs... you may have a challenge but this needs to be analyzed.


There are many benefits to move to a known, secured and well establish framework.
But moving Dolibarr to a new Framework like Laravel (for exemple) would lead to major modifications (meaning a lot of new bugs), a very huge effort to adapt the code, a stop of most of the APIs or an heavy work to modify them and create bridges so that the old ones still work, and as a result the lost of the developers and the users. So this is not an option.

I have tried to insert Laravel inside Dolibarr, but this means many modifications of Laravel Core, meaning that those changes may be swept out with a next version... Not mentioning that in this case, we have two frameworks running, so the performance are very bad. The main advantage: you do not touch anything... but this is not enough compared to the previous issues. It can be an option, but for sure not the best one.
 
The way I see it : 
-----------------
At first, building a set of **new** APIs, that will use the current database structure via Models and Controllers (the APis ones) **partly** based on the requirements of the current API, **partly** on the new framework requirements. 
This new set needs first to be well tested before going live. 

There should not have modifications at least in the 1st  version of the former APIs.
At that step, we need to be very clear on the communication for the next move. Meaning that the developers need to know what they will have to change in the future and how.

2nd step would be to adapt some of the former APIs (and continue to create the new ones) as communicated... and be responsive in term of help to the developers...

And so on until all of the main APIs are rewritten.
I guess that this process could take a few releases (years...). 


With this move we could :
- insert a new process for development that checks the quality of the code, the security & the documentation
- enhance the test process
- increase the security

The idea would be to create a professional environment:

- the daily build : based on github.com current state (full of errors :-( )
   ==> code quality and security checks (automatically)
   
- test build : based on some minor and major corrections & enhancements
   ==> code quality and security checks (automatically) + what has changed
   
- staging build (with integration / third parties) : allow the third party to test their code with the future RC
Environment where third party can use the APIs of the future version

- RC build + documentation : when almost ready for a new version

- production build + documentation

**Comment by torvista** [comment]
I posted here two years ago and it seems nothing has changed.

> I wish Eldy would comment on this...as there is a lot of energy, time, effort and experience being wasted just in this thread as it is in a vacuum.

Either
a) there is already a plan, that is not being communicated, and the developers are letting you all waste your time and effort discussing something that will be ignored.
b) they have no idea what to do and are just carrying on "putting lipstick on the the pig".
c) they are taking ideas from this thread.

This project has little coding input compared to the large user base, and so desperately needs input from anyone who shows as interest, as yourselves.
I find it depressing that the developers of this project are not actively participating in this discussion to focus it on the actual plan, whatever that is, or make one and harness the interest and skills of anyone who has bothered to post here.

Letting you all talk and not commenting is not encouraging for the future of this project, communication is the lifeblood of an open-source project.

**Comment by altairis-tof** [comment]
i may answer shortly but am afraid to be misunderstood... i try nevertheless : my point of view is that dolibar is driven by users demands.

**Comment by torvista** [comment]
> my point of view is that dolibar is driven by users demands.

Yes.

And I would say that 100% stability and no bugs in the CURRENT CODE is what **users** demand in an ERP in a business-critical application and NOT new features continually, while long-running issues are ignored.
This project seems to always be running before it can walk.

The issue of the underlying framework is of no interest to **users** until it causes them a problem/it is too slow.

The framework is of critical importance for the **future of the project** to not alienate capable coders who do not want to build on obsolete code but to encourage plugin developers to contribute **modules for new features**.
Those modules can be incorporated if necessary, but in the **future**.

Fix the existing code first, with currently 900+ issues and nearly 200 pending/ignored pull requests.
That really does not encourage anyone to contribute, why should people bother to spend their free time adding input and be ignored.
That's my opinion.

**Comment by cfoellmann** [comment]
I concur mostly with @torvista .

It is very frustrating for all the contributors to have their PRs and issues rot away.
BUT there are not enough devs to test and check and even fewer devs to merge!

The project needs a feature freeze for 1 year! 
There are many long lasting issues and unfinished features. (Example: no permission for "Variants")

And regular dev meetings in a chat where for 2 hours as many PRs and issues can be distributed to volunteers for testing.
All issues and PRs need to touched once and either closed, merged, answered or update with a note or question for the issuer.
With a lot of effort this could be cleaned up in a few months.

**Comment by altairis-tof** [comment]
so when a user wants a new development, i have to say no because the project is frozen for a year ? don't you think that by doing this, the users will go to another ERP and the developers too (because they are paid by the users...) ?

**Comment by cfoellmann** [comment]
all of the stuff that got added the last 3 versions could have been added as extensions.
The core application gets more and more bloated.

I know that the codebase is not as flexible as it must be to allow for all kinds of extensions BUT if a features needs some new hook or modification in to make it possible without breaking something else it is worth a PR.
That is the way that made WordPress bit. ANYTHING can be manipulated by hooks!!

**Comment by cfoellmann** [comment]
@eldy how about closing this thing?
There could be a new discussion but this issue is to long and outdated

**Comment by torvista** [comment]
> outdated

Only the title is out of date, the comments are still valid in 2022.

**Comment by cfoellmann** [comment]
I am pretty happy with the current course and how we can contribute.
A new discussion post referencing this issue here might be cleaner
Just my opinion

**Comment by eldy** [comment]
I agree @cfoellmann .  This ticket is long. But above all, its content is to abstract and too generic. Since most people confirm than a rewrite is not a good way, it would be better to work on concrete suggestions or contribution with example of suggestion of code change, so from other existing tickets.
